alibi_detect/od/pytorch/svm.py

```diff
@@ -0,0 +1,307 @@
+from typing import Callable, Dict, Optional, Tuple, Union
+
+import numpy as np
+import torch
+from typing_extensions import Literal, Self
+from tqdm import tqdm
+
+from alibi_detect.od.pytorch.base import TorchOutlierDetector
+from alibi_detect.utils.pytorch.losses import hinge_loss
+from alibi_detect.utils.pytorch.kernels import GaussianRBF
+
+
+class SVMTorch(TorchOutlierDetector):
+    ensemble = False
+
+    def __init__(
+        self,
+        kernel: Union['torch.nn.Module', Literal['rbf']] = 'rbf',
+        sigma: Optional[float] = None,
+        n_components: Optional[int] = None,
+        device: Optional[Union[Literal['cuda', 'gpu', 'cpu'], 'torch.device']] = None,
+    ):
+        """Pytorch backend for the Support Vector Machine (SVM) outlier detector.
+
+        Parameters
+        ----------
+        kernel
+            Used to define similarity between data points. Can be either a `torch.nn.Module` or a string. The user can
+            either pass a custom kernel as a `torch.nn.Module` or indicate a built-in kernel. The only built-in kernel
+            currently supported is the Gaussian RBF kernel. This can be specified by passing ``'rbf'`` or the user can
+            initialize the GaussianRBF kernel directly and pass it as an argument.
+        sigma
+            Kernel coefficient for the `GaussianRBF` kernel if `kernel` is ``'rbf'``. If `kernel` is a `torch.nn.Module`
+            then this argument is ignored.
+        n_components
+            Number of components in the Nystroem approximation, by default uses all of them.
+        device
+            Device type used. The default tries to use the GPU and falls back on CPU if needed. Can be specified by
+            passing either ``'cuda'``, ``'gpu'``, ``'cpu'`` or an instance of ``torch.device``.
+
+        Raises
+        ------
+        ValueError
+            If the kernel is not supported.
+        """
+        super().__init__(device=device)
+        self.n_components = n_components
+        self.kernel = kernel
+        self.sigma = sigma
+        self._init_kernel()
+
+    def _init_kernel(self) -> None:
+        """Initialize the Kernel."""
+
+        if isinstance(self.kernel, str):
+            if self.kernel not in ['rbf']:
+                raise ValueError(
+                    f'Currently only the rbf Kernel is supported for the SVM torch backend, got {self.kernel}.'
+                )
+            if self.kernel == 'rbf':
+                if self.sigma is not None:
+                    sigma = torch.tensor(self.sigma, device=self.device)
+                self.kernel = GaussianRBF(sigma=sigma)
```

Comment:
Here if `self.sigma` is `None` then `sigma` is not defined and `sigma=sigma` causes an error